src/core/drive/navigator.ts

```diff
@@ -87,8 +87,12 @@ export class Navigator {
     }
   }
 
-  formSubmissionFailedWithResponse(formSubmission: FormSubmission, fetchResponse: FetchResponse) {
-    console.error("Form submission failed", formSubmission, fetchResponse)
+  async formSubmissionFailedWithResponse(formSubmission: FormSubmission, fetchResponse: FetchResponse) {
+    const responseHTML = await fetchResponse.responseHTML
+
+    this.view.clearSnapshotCache()
+
+    this.proposeVisit(this.location, { action: "replace", response: { statusCode: fetchResponse.statusCode, responseHTML } })
```

Comment:
We don't actually want to go through the visit machinery here, since history isn't changing and we don't need the adapter to be involved. Instead we can just create a Snapshot and pass it to View#render:

```ts
const responseHTML = await fetchResponse.responseHTML
const snapshot = Snapshot.fromHTMLString(responseHTML)
this.view.render({ snapshot }, () => {})
this.view.clearSnapshotCache()
```

(The second argument to render is a callback which Visit uses to change the scroll position afterwards. Ideally we would just scroll to the top of the page here.)

Reply:
Done. Thank you for the guidance.